include/zephyr/drivers/battery.h

```diff
@@ -0,0 +1,200 @@
+	/** Temperature in 0.1 K */
+	BATTERY_TEMPERATURE,
+	/** Battery voltage (mV) */
+	BATTERY_VOLTAGE,
+
+	/** Static properties are contained in the device config */
+};
+
+#define BATTERY_INVALID_PROPERTY -ENOTSUP
+
+struct battery_get_property {
+	/** Battery property to get */
+	enum battery_property_type property_type;
+
+	/** Property field for getting */
+	union {
+		/** Fields have the format: */
+		/** BATTERY_PROPERTY_ENUM_FIELD */
+		/** type property_field; */
+
+		/** Dynamic Battery Info */
+		/** BATTERY_AVG_CURRENT, */
+		int avg_current;
+		/** BATTERY_AVG_VOLTAGE, */
+		int avg_voltage;
+		/** BATTERY_CHARGE_FET_DISABLED, */
+		bool charge_fet_disabled;
+		/** BATTERY_CURRENT, */
+		int current;
+		/** BATTERY_CUT_OFF, */
+		bool cut_off;
+		/** BATTERY_CYCLE_COUNT, */
+		uint32_t cycle_count;
+		/** BATTERY_DISCONNECT_STATE, */
+		enum battery_disconnect_state disconnect_state;
+		/** BATTERY_FLAGS, */
+		uint32_t flags;
+		/** BATTERY_FULL_CHARGE_CAPACITY, */
+		uint32_t full_charge_capacity;
+		/** BATTERY_IMBALANCE_MV, */
+		int imbalance_mv;
+		/** BATTERY_PRESENT_STATE, */
+		enum battery_present battery_present_state;
+		/** BATTERY_REMAINING_CAPACITY, */
+		int remaining_capacity;
+		/** BATTERY_RUN_TIME_TO_EMPTY, */
+		uint32_t run_time_to_empty;
+		/** BATTERY_STATE_OF_CHARGE, */
+		uint8_t state_of_charge;
+		/** BATTERY_TEMPERATURE, */
+		int temperature;
+		/** BATTERY_VOLTAGE, */
+		int voltage;
+	};
+};
+
+/**
+ * Cut off charge from battery
+ *
+ * @return non-zero if the battery driver doesn't support cut off.
+ */
+typedef int (*battery_cut_off)(const struct device *dev);
+
+/**
+ * @brief Get a battery property
+ *
+ * @param dev Pointer to the battery device
+ * @param prop pointer to battery_get_property struct where the property struct
+ * field is set by the caller to determine what property is read from the
+ * battery device.
+ *
+ * @return 0 if successful, negative errno code if failure.
+ */
+typedef int (*battery_get_property_t)(const struct device *dev, struct battery_get_property *prop);
+
+/**
+ * Read time in seconds left when discharging.
+ *
+ * @param capacity	Destination for remaining time in seconds.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_to_empty)(const struct device *dev, int *seconds);
+
+/**
+ * Read time in seconds left to full capacity when charging.
+ *
+ * @param capacity	Destination for remaining time in seconds.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_to_full)(const struct device *dev, int *seconds);
+
+/**
+ * Calculate battery time in seconds, under an assumed current.
+ *
+ * @param rate		Current to use for calculation, in mA.
+ *			If > 0, calculates charging time; if < 0, calculates
+ *			discharging time; 0 is invalid and sets seconds=0.
+ * @param seconds	Destination for calculated time in seconds.
+ * @return non-zero if error.
+ */
+typedef int (*battery_time_at_rate)(const struct device *dev, int rate, int *seconds);
+
```

Comment:
If this function simply calculates time to empty from the battery capacity, then this function should not be an API call, it should be an inline helper, and simply return the time in seconds, without the out param.
```
static inline int (*battery_calculate_time_to_empty)(int capacity, int rate)
{
    ...
    ...
    return time_remaining;
}
```